opensource/isdn4k-utils/divertctrl/divertctrl.man
.\" $Id: divertctrl.man.in,v 1.4 2001/06/10 17:38:25 werner Exp $
.\"
.\" CHECKIN $Date: 2001/06/10 17:38:25 $
.\"
.\" Process this file with
.\" groff -man -Tascii divertctrl.8 for ASCII output, or
.\" groff -man -Tps divertctrl.8 for PostScript output
.\"
.TH divertctrl 8 "2001/06/10" isdn4k-utils-3.2p1 "Linux System Administration"
.SH NAME
divertctrl \- set/query ISDN diversion services for (E)DSS1 protocol 
.SH SYNOPSIS
.BI divertctrl " [wait] command driverid " ...
.SH DESCRIPTION
.B divertctrl
is used to de/activate call diversions and query actual activated diversion rules. 
The i4l diversion services only work using the (E)DSS1 D-channel protocol in conjunction with the HiSax passive card driver. For using the services the global isdn drivers need to be compiled with support for the diversion services.
Additionally the dss1_divert module has to be loaded. This module doesn't require nor support any parameters at load time. 
After successfully loading the module an entry /proc/net/isdn/divert should appear in the filesystem. When called without any parameters the divertctrl program outputs a short help screen. Otherwise the first parameter needs to be a command followed by a valid driver id. The command may be preceeded by the optional wait keyword specifying the program to wait until the desired command could be completed or failed returning the result via the exitcode. Otherwise the programm immediately returns after invoking the desired action which may not be completed at this moment. 
 For some commands the value "-" may be used as a valid driver id specifying all available drivers. The driver id is equivalent to the id parameter specified when loading the HiSax driver for a particular card.
All further parameters are command dependant.    
The divertctrl program may only be used with root access for security reasons.
.LP 
The diversion services for i4l may be used in two independant ways:
.HP
.SH 1. Static call diversion   
First possibility to handle diversions of incoming calls is to use static diversion
provided inside the providers exchange.
A static diversion once activated inside the providers exchange requires no interaction with i4l. The machine may even be shut down, but the diversion keeps active until it is explicitely deactivated.     
The divertctrl tool allows to set/reset and query such static rules if the service is supported and has been subscribed at the providers side.
This services are only available in some countries like germany. In other countries (like the netherlands) keypad control is used to de/activate such static rules.    
Static rules may be set/reset and queried independantly by MSN (multiple subscriber number), basic service (telephony, digital data, ..) and diversion procedure. Three diversion procedures are defined in the ETSI specs and may be used with the i4l diversion services:
.LP
.B CFU (call forward unconditional) 
is a procedure diverting all incoming calls unconditionally for the programmed MSN and basic service. The call will never be announced at your side until CFU is deactivated again. 
.LP
.B CFNR (call forward not reachable) 
is a procedure diverting all incoming calls for the programmed MSN and basic service after locally signalling and waiting a certain timeout period.
If the call is not answered during this timeout period it will be diverted to the new destination. The timeout period is fixed in the providers exchange and is normally 3 rings (about 12 to 15 seconds).
.LP
.B CFB (call forward busy) 
is a procedure diverting all incoming calls for the programmed MSN and basic service when all local resources taking the call are exhausted and busy.
.HP
.SH Commands for handling static call diversions
.B activate 
driverid <cfu,cfnr,cfb> msn service destination
.LP
Activate a static diversion for the given driver, msn and service diverting the call to the specified destination. All parameters need to be supplied, no wildcards are allowed. Only one of the three diversion procedures cfu, cfnr, cfb must be supplied.
The value for the service may be taken from the table of numeric codes of basic services. The value 0 specifies all available/subscribed services.      
.LP
.B deactivate 
driverid <cfu,cfnr,cfb> msn service
.LP
Deactivate a static diversion for the given driver, msn and service. All parameters need to be supplied, no wildcards are allowed. Only one of the three diversion procedures cfu, cfnr, cfb must be supplied.
The value for the service may be taken from the table of numeric codes of basic services. The value 0 specifies all available/subscribed services.      
.LP
.B interrogate 
driverid <cfu,cfnr,cfb> [msn] [service]
.LP 
Query static diversions for the given driver, msn and service. Only one of the three diversion procedures cfu, cfnr, cfb must be supplied. The msn and service parameters are optional.
The value for the service may be taken from the table of numeric codes of basic services. The value 0 specifies all available/subscribed services.      
If msn and/or service parameters are not specified all matching diversions are reported via stdout. But it is advisable always to specify all parameters to keep the list as short as possible. All known providers exchanges refuse to return diversion lists longer than 256 bytes. In this case an empty response is generated by the exchange even if there are diversions active ! 
.HP
.SH 2. Dynamic call diversion   
Additionally the i4l diversion services offer a more flexible possibility to control call forwarding. Using the dynamic call diversion the user has the possibilty to specify rules for call forwarding by additional criterias. The reaction to an incoming call may be dependant of MSN, basic service, caller id, local subaddress, caller subaddress and local resource (busy) state. The parameters may be specified with wildcards, so that call criterias may be grouped to match.
Additionally the diversion actions may be supplied with a precise timeout value which is not dependant on any providers defaults. In order to work, the supplumentary service CD (call deflection) has to be available and subscribed at the providers exchange. 
The dynamic diversion services are fully handled inside your machine, so it must be powered up and activated for the required purpose. After a successfull dynamic diversion (so called deflection) no local line resources are required. The lines are free for further incoming calls.
.LP
Dynamic Call deflection is controlled by a rule chain the user has to supply using the divertctrl program. When an incoming call arrives, calling data is compared against the rules in the chain. If an incoming call matches a rule, this rule is taken to execute the desired action. All following rules are ignored. 
If there is no rule match the diversion services simply ignore the call.        
.HP
.SH Commands for handling dynamic call diversions
.B flushrules
[driverid]
.LP
Flushes (deletes) all rules for the selected driver. If no driverid is given or it is specified as wilcard - all rules for all drivers are removed.
It is advisable to call this command first when a complete new ruleset is to be generated, to avoid conflicts with previous set rules.  
.LP
.B appendrule
driverid action msn si1 si2 callerid screen delay option destnumber
.LP 
This command appends a single rule at the end of the existing rule chain.
If the call arrives through the desired driver, addresses the selected msn, si1, si2 and matches the desired callerid and option the specified action is executed.
A value of - may be specified for the driverid to match the rule for all available drivers. The msn may be specified with a trailing - wildcard. 
for example the value 123 only matches an incoming call to msn 123, but specifying 123- matches all msn starting with 123 followed by any digits or subaddresses which will not verified. If only - is specified the rule matches all msn and subaddresses.
If your isdn line supports subaddressing it is advisable to terminate all msn values with a - because the msn check includes a possibly available subadress which then may be reported as 123.456 for msn 123 with subaddresses 456 for example.
Subaddressing is a special DSS1 feature not available in most countries and normally needs to be specially subscribed. So most people need not to think about it.
The value of si1 represents the numeric code of the desired and checked basic service for the incoming call. This value may be selected from the table below or just the value 0 specified for all services to match.
The value of si2 represents an additional service indicator for high layer compatibilty and is only included for completeness. Just set it to 0 at the moment.
The callerid must match the number of the caller including the subaddress if available. Again the special wildcard - may be used to match specific groups of numbers. 
Additionally a simple value of 0 may be specified. In this case the rule will match only calls coming in without a caller indentification. This will be the case if the caller originates from a network not supporting callerids or the caller suppressed the identification.
The option parameter may take the values 0 to 2 and specifies whether the rule applies only during special local busy states.
The value of 0 lets the rule be valid during any local busy state. A value of 1 lets the rule only apply to incoming calls if the call is in a non waiting state.
A value of 2 applies te rule only to such calls which arrive as waiting. This is normally the case when all local resources (B-channels) are already in use.
If the rule criterias mentioned before match the incoming call, no following rules will be checked and the desired action will be executed. The value for the parameter action is numeric and may take the values 0 to 5 at the moment.
A value of 0 lets the call to be ignored. The call will not be reported through the ascii interface and not checked against any following rules.
A value of 1 will report the call through the ascii interface but no other action will take place. If the value 2 is specified the call will be reported through the ascii interface and actively put in a local proceeding state.
This means that the providers exchange is told, that your side needs more time to check whether the call may be handled and in which way this will be done.
This value only is intended for use with local or remote client software watching the ascii interface and deciding what to do. No ringing signal is send to the caller until the decision has been made or a timeout (typically 5 to 15 seconds) occurs.
An example would be a software which announces the call to a user and requests the desired action. At the moment a client software is under development, but still not available, so this value may only be intersting for programmers which want to write their own client.
If value of 4 is specified the call will be actively rejected.
The value of 5 is not primary an diversion function and allows an i4l
network device to be started for dialing out when the rule
matches. The destination number parameter specifies the network device
(for example ippp0) to e dialed. The incoming call itself is not accepted. 
The values from 0-2 and 4 don't require a destination number to be specified, as the incoming call will not be deflected in this cases.
The last, but most interesting value for most people will be the value 3. Specifying it, will let the call to be deflected/diverted actively.
For this reason additional parameters are taken for interpretation. First of all destnumber specifies the final number the call should be diverted to.
The parameter delay specifies after how many seconds the call will be diverted towards the new destination. A value of 0 deflects the call immediately like the cfu in static diversons, any other value first announces the caller a ringing state until the time is elapsed and then the call will be diverted like in static cfnr.
During the ringing phase every other device on your line may pick up the call of course.          
The value of the parameter screen may take the values 0 to 2 and specifies if the diversion is presented to the caller. A value of 0 denies to show the caller that and where the call has been deflected. Specifying a value of 1 only shows that the call has been diverted but doesn't show to which final destination this will happen. A value of 2 lets the caller know all information of the diversion (fact of diversion and number diverted to).
.LP
.B insertrule
driverid action msn si1 si2 callerid screen delay option destnumber
.LP 
This command inserts a single rule at the beginning before the first already existing rule in the chain. 
All parameters and descriptions are the same as for the appendrule command.

.LP
.SH Numeric codes of basic services
  0  all services
  1  speech
  2  unrestricted digital information 
  3  audio 3.1 kHz
  4  unrestricted digital info with tone announcements
  5  multirate
 32  telephony 3.1 kHz
 33  teletex
 34  telefax group 4 class 1
 35  videotex syntax based
 36  videotelephony
 37  telefax group 2/3
 38  telephony 7 kHz
 39  eurofiletransfer
 40  filetransfer and access management
 41  videoconference
 42  audio graphic conference

When diversion of speech calls is desired at least services 1, 2 and 32 should be specified.


.LP
.SH Interfacing to other programs
The /proc/net/isdn/divert device may be used for debug purposes or interfacing the diversion services to other programs. It may be multiple opened. All operations as well as incoming calls may be watched reading the ascii output of the interface. One possible application would be a remote client announcing and logging incoming calls and diversion actions inside the local network. Such logging service could be invoked via inetd.    
      
.LP
.SH BUGS
With some commands an explicit driverid needs to be specified under certain conditions even if wildcards should be allowed. If you get a core dump using wildcards try to use a cmd line specifying a single interface. 
This man page is still not complete.

.LP
.SH AUTHOR
Werner Cornelius <werner@isdn4linux.de or werner@isdn-development.de>





opensource/isdn4k-utils/divertctrl/divertctrl.c
/* 
 * $Id: divertctrl.c,v 1.5 2002/07/04 10:36:54 paul Exp $
 *
 * Control program for the dss1 diversion supplementary services. (User side)
 *
 * Copyright 1999       by Werner Cornelius (werner@ikt.de)
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA. 
 *
 * $Log: divertctrl.c,v $
 * Revision 1.5  2002/07/04 10:36:54  paul
 * Include stdlib.h for prototypes;
 * Make some things depend on NETWORK_DIAL define.
 *
 * Revision 1.4  2001/06/10 17:38:25  werner
 *
 * Updated tool and manpage with new remote dial feature
 *
 * Revision 1.3  2001/01/09 19:27:55  werner
 *
 * Added Output for interrogate and description for services.
 * Thanks for help from Wolfram Joost.
 *
 * Revision 1.2  1999/09/02 13:24:14  paul
 * fixed some compile warnings
 *
 * Revision 1.1  1999/05/07 21:33:01  werner
 * Initial release of divertctrl
 *
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include "isdn_divert.h"

static char number_types[7] = {
       'U',
       'I',
       'N',
       'S',
       'M',
       '?',
       'A' };

static char cf_types[4][5] = {
        "CFU", "CFB", "CFNR", "?" };

static struct {
       unsigned   type;
       char       *name;
} service_types[18] = {
       {  0, "all services" },
       {  1, "speech" },
       {  2, "unrestricted digital information" },
       {  3, "audio 3.1 kHz" },
       {  4, "unrestricted digital information wtaa" },
       {  5, "multirate" },
       { 32, "telephony 3.1 kHz" },
       { 33, "teletex" },
       { 34, "telefax group 4 class 1" },
       { 35, "videotex syntax based" },
       { 36, "videotelephony" },
       { 37, "telefax group 2/3" },
       { 38, "telephony 7 kHz" },
       { 39, "eurofiletransfer" },
       { 40, "filetransfer and access management" },
       { 41, "videoconference" },
       { 42, "audio graphic conference" } };

char *progname; /* name of program */ 
ulong  drvid; /* ids of driver */
int  fd; /* file descriptor for read and ioctl */
divert_ioctl dioctl; /* structure exchanging ioctl data */
char **argp; /* pointer to actual parameter */
int argrest; /* remaining args */
int waitreq; /* wait requested */
int exitcode;/* exitcode for program */

/****************************/
/* print a short usage help */
/****************************/
void usage(void)
{ close(fd);
  fprintf(stderr,"usage: %s [wait] <Command> [parms]\n\n",progname);
  fprintf(stderr,"[wait]    : wait optional keyword until -> wait until command executed \n");
  fprintf(stderr,"<Command> : desired command to execute\n");
  fprintf(stderr,"   activate    <drivers> <cfu,cfb,cfnr> <msn> <service/0> <dest_no>\n");  
  fprintf(stderr,"   deactivate  <drivers> <cfu,cfb,cfnr> <msn> <service/0> \n");  
  fprintf(stderr,"   interrogate <drivers> <cfu,cfb,cfnr> <msn> <service/0> \n");  
  fprintf(stderr,"   listrules   <drivers> -> list all rules for drivers\n");
  fprintf(stderr,"   ignore      <callid> [uus1_string] -> ignore call with callid/send uus1\n");
  fprintf(stderr,"   alert       <callid> [uus1_string] -> alert call with callid/send uus1\n");
  fprintf(stderr,"   deflect     <callid> [to_nr] -> deflect call with callid to to_nr \n");
  fprintf(stderr,"   appendrule  <drivers> -> append a rule at list end\n");
  fprintf(stderr,"   insertrule  <drivers> -> insert a rule at list start\n");
  fprintf(stderr,"   flushrules  [drivers] -> flush all rules for specified drivers\n");
  fprintf(stderr,"   append- and insertrule take the following arguments:\n");
#ifdef NETWORK_DIAL
  fprintf(stderr,"   <drivers> <action> [msn] [si1] [si2] [caller] [screen] [delay] [callopt] [destnr/if]\n");
#else
  fprintf(stderr,"   <drivers> <action> [msn] [si1] [si2] [caller] [screen] [delay] [callopt] [destnr]\n");
#endif
  fprintf(stderr,"   drivers -> driver names for D-chans, %% separated, - = all D-chan\n"); 
  fprintf(stderr,"   action -> 0/ignore 1/report 2/proceed 3/alert 4/reject\n");
  fprintf(stderr,"   msn -> - = wildcard or explicit selected msn.subaddress\n");
  fprintf(stderr,"   when action = 5 then if specifies a network interface to dial\n");
  fprintf(stderr,"   si1 -> 0 = all services, else bitmask for services 1..7\n");
  fprintf(stderr,"   si2 -> 0 = ignore, else specified value\n");
  fprintf(stderr,"   caller -> - = wildcard, 0 = unknown, else number.subaddress\n");
  fprintf(stderr,"             a - trailing number digits is handled as wildcard\n");
  fprintf(stderr,"   screen -> 0 = noscreen 1 = divertonly 2 = divertto\n");
  fprintf(stderr,"   delay -> time in seconds until hangup or divert in auto mode\n");
  fprintf(stderr,"   callopt -> 0 = all calls, 1 = only non waiting, 2 = only waiting\n");
  fprintf(stderr,"   destnr -> destination number and subaddress\n");
  fprintf(stderr,"   defaults: allmsn, si1=0, si2=0, all callers, noscreen, delay=10\n");
  fprintf(stderr,"\n");  
  exit(-1);
} /* usage */


/******************************************/
/* get the driver parameter and set drvid */
/******************************************/
void getdrvid(void)
{ char *p,*p1;

  argp++; argrest--;

  if (argrest < 1) 
   { drvid = DEFLECT_ALL_IDS;
     return;
   }

  if ((*argp)[0] == '-')
    drvid = DEFLECT_ALL_IDS;
   else
    { drvid = 0; /* first no device */
      p = *argp; 
      while (p)
       { p1 = p;
         p = strchr(p1,'%');
         if (p) *p++ = '\0'; 
         strcpy(dioctl.getid.drvnam,p1); /* set string to query */
         if (ioctl(fd,IIOCGETDRV,&dioctl))
          { fprintf(stderr,"driver %s not found \n",p1);
            close(fd);
            exit(-1);
          } 
         drvid |= (1L << dioctl.getid.drvid);
       }
    }
  argp++; argrest--;
} /* getdrvid */ 

/***********************************/
/* set the rule pars from cmd line */
/***********************************/
void setrulepar(void)
{ divert_rule *dr = &dioctl.getsetrule.rule;
  unsigned u;

  /* first set the defaults */
  dr->drvid = 0; /* no driver */
  dr->action= DEFLECT_IGNORE; 
  strcpy(dr->my_msn,"0");
  dr->si1   = 0;
  dr->si2   = 0;
  strcpy(dr->caller,"-");
  dr->screen= 0; /* no screening */
  dr->waittime = 0;
  dr->callopt = 0;
  dr->to_nr[0] = '\0';

  if (!argrest) return; 

  /* action */
  if ((sscanf(*(argp),"%u",&u) <= 0) || (u > 5))
   { fprintf(stderr,"invalid action value %s\n",*argp);
     usage();
   }
  dr->action = u & 0xFF;
#ifdef NETWORK_DIAL
  if (((dr->action == DEFLECT_ALERT) || (dr->action == NETWORK_DIAL)) && (argrest != 9))
#else
  if ((dr->action == DEFLECT_ALERT) && (argrest != 9))
#endif
   { fprintf(stderr,"alerting/dialing action must be supplied with all parms\n");
     usage();
   }    
  if (!--argrest) return;

  /* copy msn */
  strcpy(dr->my_msn, *(++argp));
  if (!--argrest) return; 

  /* si1 */
  if (sscanf(*(++argp),"%u",&u) <= 0)
   { fprintf(stderr,"invalid si1 value %s\n",*argp);
     usage();
   }
  dr->si1 = u & 0xFF;   
  if (!--argrest) return;
  
  /* si2 */
  if (sscanf(*(++argp),"%u",&u) <= 0)
   { fprintf(stderr,"invalid si2 value %s\n",*argp);
     usage();
   }
  dr->si2 = u & 0xFF;   
  if (!--argrest) return;
  
  /* caller */
  strcpy(dr->caller,*(++argp));
  if (!--argrest) return;
  
  /* screen */
  if ((sscanf(*(++argp),"%u",&u) <= 0) || (u > 2))
   { fprintf(stderr,"invalid screen value %s\n",*argp);
     usage();
   }
  dr->screen = u & 0xFF;   
  if (!--argrest) return;

  /* delay */
  if ((sscanf(*(++argp),"%u",&u) <= 0) || (u > 99))
   { fprintf(stderr,"invalid delay value %s\n",*argp);
     usage();
   }
  dr->waittime = u & 0xFF;   
  if (!--argrest) return;

  /* callopt */
  if ((sscanf(*(++argp),"%u",&u) <= 0) || (u > 2))
   { fprintf(stderr,"invalid callopt value %s\n",*argp);
     usage();
   }
  dr->callopt = u & 0xFF;   
  if (!--argrest) return;

  /* to_nr */
  strcpy(dr->to_nr,*(++argp));
  if (!--argrest) return;  
  
  fprintf(stderr,"to many cmd line parms\n");
  usage();
} /* setrulepar */

/************************************************/
/* getcallid fetches the callid for id commands */
/************************************************/
void getcallid(void)
{
  argp++; argrest--;

  if (argrest < 1) 
   { fprintf(stderr,"callid missing\n");
     usage();  
   }

  if (sscanf(*argp,"%lu",&dioctl.fwd_ctrl.callid) != 1)
    { fprintf(stderr,"invalid callid value %s\n",*argp);
      usage();
    }
  argp++; argrest--;
} /* getcallid */

/***************************************************************************/
/* getcfpar fetches the parms for de/activating and interrogating services */
/***************************************************************************/
#define CFP dioctl.cf_ctrl
void getcfpar(void)
{ unsigned u;

  /* first set the defaults */
  CFP.service = 0; /* all services */
  CFP.msn[0] = '\0'; /* all msns */
  CFP.fwd_nr[0] = '\0';
  
  if (!argrest)
   { fprintf(stderr,"no procedure specified \n");
     usage();
   } 
  if (!strcmp(*argp,"cfu")) 
    CFP.cfproc = 0;
     else 
      if (!strcmp(*argp,"cfb"))
        CFP.cfproc = 1;
         else
          if (!strcmp(*argp,"cfnr"))
            CFP.cfproc = 2;
             else
	      { fprintf(stderr,"invalid procedure %s\n",*argp);
                usage();
              }
  if (waitreq) CFP.cfproc |= 0x80;
  argp++; argrest--; 

  strcpy(CFP.msn,*(argp++));
  if (!(--argrest)) return;

  if ((sscanf(*argp,"%u",&u) != 1) ||
      (u > 255))
   { fprintf(stderr,"invalid service [0-255] %s\n",*argp);
     usage();
   }    
  CFP.service = u & 0xFF;
  argrest--; argp++;
} /* getcfpar */ 
#undef CFP

/********************/
/* execute commands */
/********************/
static void do_command(void)
{ divert_rule *dr = &dioctl.getsetrule.rule;
  divert_ioctl ioc;
  int i;
  ulong u,cmd;
  fd_set rdfs;
  struct timeval tv;
  char read_buf[512];
  ssize_t read_len;
  char *read_ptr;
  ulong read_procid;
  ulong read_counter;
  char read_driver[30];
  u_long read_msn_type;
  char read_msn[35];
  u_long read_cfproc;
  u_long read_service;
  u_long read_tonr_type;
  char read_tonr[35];
  int read_char_count;
  char read_end;
  char read_found;

  if (!strcmp(*argp,"listrules"))
    { getdrvid();
      dioctl.getsetrule.ruleidx = 0; /* get first rule */ 
      while (1)
       { /* fetch all rules */
         if (ioctl(fd,IIOCGETRULE,&dioctl)) return; /* no more rules */
         i = 0; /* first interface */
         dr->drvid = drvid & dr->drvid; /* only requested bits */ 
         while (dr->drvid)  
	  { if (!(dr->drvid & 1))
             { dr->drvid >>= 1; /* next entry */
               i++;
               continue;
             } 
            if (dr->drvid == DEFLECT_ALL_IDS)
             { printf("%10s","-"); /* all if */
               dr->drvid = 0; /* no further output */
             }  
             else
              { ioc.getid.drvid = i++; /* set para */
                dr->drvid >>= 1;
                if (ioctl(fd,IIOCGETNAM,&ioc)) continue; /* invalid driver */
                if (!ioc.getid.drvnam[0]) continue; /* no name */
                printf("%10s ",ioc.getid.drvnam);
              }
            printf("%15s %20s %2d %3d %d %2d %d %s\n",
                   dr->my_msn,  /* msn */ 
                   dr->caller,  /* caller id */
                   dr->si1,     /* si1 0 = all */
                   dr->si2,     /* si2 0 = all */
                   dr->action,  /* desired action */
                   dr->waittime,/* wait until action */
                   dr->callopt, /* option for waiting calls */
                   dr->to_nr);  /* deflect to */
          } /* interface match */ 
         dioctl.getsetrule.ruleidx++; /* next rule */ 
       } /* fetch all rules */
    } /* listrules */
  else
  if (!strcmp(*argp,"deflect"))
    { getcallid();
      dioctl.fwd_ctrl.subcmd = 2;
      if (argrest)
       { strcpy(dioctl.fwd_ctrl.to_nr,*argp);
         argp++; argrest--;
       }
      else
         dioctl.fwd_ctrl.to_nr[0] = '\0';
      exitcode = ioctl(fd,IIOCDODFACT,&dioctl);
    } /* deflect */
  else
  if (!strcmp(*argp,"alert"))
    { getcallid();
      dioctl.fwd_ctrl.subcmd = 1;
      if (argrest)
       { strcpy(dioctl.fwd_ctrl.to_nr,*argp);
         argp++; argrest--;
       }
      else
         dioctl.fwd_ctrl.to_nr[0] = '\0';
      exitcode = ioctl(fd,IIOCDODFACT,&dioctl);
    } /* alert */
  else
  if (!strcmp(*argp,"ignore"))
    { getcallid();
      dioctl.fwd_ctrl.subcmd = 0;
      if (argrest)
       { strcpy(dioctl.fwd_ctrl.to_nr,*argp);
         argp++; argrest--;
       }
      else
         dioctl.fwd_ctrl.to_nr[0] = '\0';
      exitcode = ioctl(fd,IIOCDODFACT,&dioctl);
    } /* ignore */
  else
  if (!strcmp(*argp,"insertrule"))
    { getdrvid();
      setrulepar();
      dioctl.getsetrule.ruleidx = 0; /* start of list */
      dr->drvid = drvid; /* driver id */
      ioctl(fd,IIOCINSRULE,&dioctl);
    } /* insertrule */
  else
  if (!strcmp(*argp,"appendrule"))
    { getdrvid();
      setrulepar();
      dioctl.getsetrule.ruleidx = -1; /* end of list */
      dr->drvid = drvid; /* driver id */
      ioctl(fd,IIOCINSRULE,&dioctl);
    } /* appendrule */
  else
  if (!strcmp(*argp,"flushrules"))
    { getdrvid(); 
      dioctl.getsetrule.ruleidx = 0; /* start of list */
      while (1)
       { /* fetch all rules */
         if (ioctl(fd,IIOCGETRULE,&dioctl)) return; /* no more rules */
         if (drvid & dr->drvid)
	  { dr->drvid &= ~drvid; /* delete drivers */
            if (dr->drvid)
	     { ioctl(fd,IIOCMODRULE,&dioctl);
               dioctl.getsetrule.ruleidx++;
             }       
            else
             ioctl(fd,IIOCDELRULE,&dioctl); 
          } 
         else
          dioctl.getsetrule.ruleidx++;
       } /* fetch all rules */
    } /* flushrules */
  else
  if ((!strcmp(*argp,"activate")) || (!strcmp(*argp,"deactivate"))
      || (!strcmp(*argp,"interrogate")))
    { if (*argp[0] == 'a') 
        cmd = IIOCDOCFACT;
       else if (*argp[0] == 'i')
         cmd = IIOCDOCFINT; 
       else
         cmd = IIOCDOCFDIS; 
      getdrvid();
      getcfpar(); 
      dioctl.cf_ctrl.fwd_nr[0] = '\0';
      if (cmd == IIOCDOCFACT)
       { if (!argrest)
          { fprintf(stderr,"missing diverted to number \n");
            usage();
          }
         strcpy(dioctl.cf_ctrl.fwd_nr,*(argp++));
         argrest--;
       } /* only activate */
      dioctl.cf_ctrl.drvid = 0;
      u = drvid; /* list of drivers */
      while (u)
       { if (u & 1)
	 { if ((exitcode = ioctl(fd,cmd,&dioctl)) > 0) 
              return; /* error executing diversion */
            else
             if ((exitcode < 0) && (drvid != DEFLECT_ALL_IDS))
               return;
                        read_end = 0;
            read_found = 0;
            while ( !read_end && (cmd == IIOCDOCFINT) )
             { FD_ZERO(&rdfs);
               FD_SET(fd,&rdfs);
               tv.tv_sec = 4;
               tv.tv_usec = 0;
               if (select(fd+1,&rdfs,NULL,NULL,&tv) != 1)
                { break;
                }
               read_len = read(fd,read_buf,sizeof(read_buf)-1);
               if (read_len)
               { read_buf[read_len] = 0;
                 read_ptr = read_buf;
                 while (*read_ptr)
                  { if ( ((*read_ptr != (DIVERT_REPORT | 0x30)) ||
                        (read_ptr[1] != ' ')) && strncmp(read_ptr,"128 ",4) )
                    { while (*read_ptr && (*read_ptr != '\r'))
                        read_ptr++;
                      if (*read_ptr)
                        read_ptr++;
                      continue;
                    }
                    if ( !strncmp(read_ptr,"128 ",4)) {
                       sscanf(read_ptr + 4,"%lx",&read_procid);
                       if (read_procid == dioctl.cf_ctrl.procid) {
                          puts("Error. :-(");
                          read_end = 1;
                          break;
                       }
                    }
                    else {
                       read_ptr += 2;
                       sscanf(read_ptr,"%lx %lu %s %lu %s %lx %lu %n",
                              &read_procid,&read_counter,read_driver,
                              &read_msn_type,read_msn,&read_service,&read_cfproc,
                              &read_char_count);
                       read_ptr += read_char_count;
                       read_end = (read_procid == dioctl.cf_ctrl.procid) && !read_counter;
                       if (read_end)
                        { if (!read_found)
                            puts("Nothing active.");
                          break;
                        }
                       if ( read_procid == dioctl.cf_ctrl.procid )
                        { sscanf(read_ptr,"%lu %s",&read_tonr_type,read_tonr);
                          if (read_msn_type > 6)
                             read_msn_type = 5;
                          if (read_tonr_type > 6)
                             read_tonr_type = 5;
                          if (read_cfproc > 3)
                             read_cfproc = 3;
                          read_found = 1;
                          printf("%-8s %c %-8s %-4s %c %-10s ",read_driver,
                                 number_types[read_msn_type],read_msn,
                                 cf_types[read_cfproc],
                                 number_types[read_tonr_type],read_tonr);
                          for (read_char_count = 0; read_char_count < 18; read_char_count++)
                           { if (service_types[read_char_count].type == read_service)
                              { puts(service_types[read_char_count].name);
                                break;
                              }
                           }
                          if (read_char_count == 18)
                             puts("unkown");
                        }
                     }
                     while (*read_ptr && (*read_ptr != '\r'))
                        read_ptr++;
                     if (*read_ptr)
                        read_ptr++;
                 }
               }
             }
            exitcode = 0;
          }
         u >>= 1; /* next driver */  
         dioctl.cf_ctrl.drvid++;    
       } /* for all drivers */
    } /* activate */
  else
    { fprintf(stderr,"unknown command: %s\n",*argp);
      usage();
    }
} /* do_command */


int main(int argc, char *argv[])
{

  exitcode = 0; /* default no error */

  /* first get our own program name */
  if ((progname = strrchr(argv[0], '/')))
    progname++;
  else
    progname = argv[0];

  /* now open the interface */
  if ((fd = open("/proc/net/isdn/divert", O_RDWR)) < 0)
    { if ((fd = open("/dev/isdndivert", O_RDWR)) < 0)
        { fprintf(stderr,"error opening device (dss1_divert module not loaded ?)\n");
          exit(-1);
        } 
    } 
  
  /* check the interface version */
  dioctl.drv_version  = -1; /* invalidate */
  if (ioctl(fd,IIOCGETVER,&dioctl))
    { fprintf(stderr,"could not execute ioctl call \n");
      close(fd);
      exit(-1);
    }  
  if ( dioctl.drv_version != DIVERT_IIOC_VERSION)
    { fprintf(stderr,"versions of %s and driver incompatible\n",progname);
      close(fd);
      exit(-1);
    } 

  argrest = argc - 1; /* remaining number of parameters */
  argp = argv + 1; /* pointer to next parameter/command */

  waitreq = 0; /* if no wait desired */
  if (argrest >= 1)
   if (!strcmp(*argp,"wait"))
    { waitreq = 1; 
      argp++;
      argrest--;
    }  
  /* a minimum of 1 additional cmd line parameters is needed */
  if (argrest < 1) usage();
  
  do_command(); 
  close(fd);
  exit(exitcode);
} /* main */


